backend/agents/types.py

```diff
@@ -0,0 +1,24 @@
+from typing import Callable, Protocol, Tuple, TypeVar

```

Comment:
I wasn't aware Python had this `lib/typing` module. I've probably used it under the hood a bunch of times but not directly pulled it into a module like this. Neat!